content/6_Advanced/Offline_Del.mdx

```diff
@@ -1,86 +1,279 @@
 ---
 id: offline-del
 title: 'Offline Deletion'
-author: Benjamin Qi, Siyong Huang
+author: Benjamin Qi, Siyong Huang, Chongtian Ma
 prerequisites:
   - dsu
 description: 'Erasing from non-amortized insert-only data structures.'
 frequency: 0
 ---
 
-## Offline Deleting from a Data Structure
-
 Using a persistent data structure or rollbacking, you are able to simulate
 deleting from a data structure while only using insertion operations.
 
-<Resources>
-	<Resource
-		source="CP-Algorithms"
-		title="Deleting from a data structure in O(T(n) log n)"
-		url="https://cp-algorithms.com/data_structures/deleting_in_log_n.html"
-		starred
-	>
-		includes code (but no explanation) for dynamic connectivity
-	</Resource>
-</Resources>
+## DSU With Rollback
+
+**DSU with Rollback** is an extension to DSU that saves merges and can undo the previous $k$ merges.
+
+<Problems problems="rollback" />
+
+<Warning title="Watch Out!">
+
+Because path compression is amortized, it does not guarantee 
+$\mathcal{O}(N \log^2 N)$ runtime. You _must_ use merging by rank. An explanation is given [here](https://codeforces.com/blog/entry/90340?#comment-787571).
+
+</Warning>
+
+### Implementation
+
+Adding on to usual DSU, we can  store the parent 
+and sizes of nodes that are being merged before each merge. This allows us revert each node to their parents before the merge so that the rollback function can use the information to undo the merges.
+
+We can store each state of the DSU using an integer, captured by the snapshot function which returns the number of old merges that has not been rolled back. It's similar taking a picture of something, and years later going back into your photo album and scrolling up until you find this picture.
```

Comment:
maybe add an example of what this might be
just like 4 nodes or smth

Reply:
huh how do draw this out

Reply:
that's for you to figure out man idk
draw it by hand like ryan does
use the 3blue1brown thing like brebenel does
use networkx
so many options

Reply:
this isnt something u can draw out though..
its just pushing into a vector and undoing the last thing on the vector

Reply:
heaven forbid you write out a list of numbers